src/Graphics/Widgets/Widget.cc

```diff
@@ -61,28 +61,59 @@ void WidgetBase::setPosition(const Point& p)
   position_ = p;
 }
 
-void WidgetBase::setToScale(const Vector& flipAxis)
+void WidgetBase::setToScaleAxis(MouseButton btn, const Vector& flipAxis, glm::mat4 scaleTrans, int scaleAxisIndex)
 {
-  movementType_ = WidgetMovement::SCALE;
+  if(btn == MouseButton::NONE) return;
+  movementType_[btn] = WidgetMovement::SCALE_AXIS;
   flipAxis_ = glm::vec3(flipAxis.x(), flipAxis.y(), flipAxis.z());
+  scaleAxisIndex_ = scaleAxisIndex;
+  scaleTrans_ = scaleTrans;
 }
 
-void WidgetBase::setToRotate()
+void WidgetBase::setToScaleAxisHalf(MouseButton btn, const Vector &flipAxis,
+                                glm::mat4 scaleTrans, int scaleAxisIndex) {
+  if (btn == MouseButton::NONE)
+    return;
+  movementType_[btn] = WidgetMovement::SCALE_AXIS;
+  flipAxis_ = glm::vec3(flipAxis.x(), flipAxis.y(), flipAxis.z());
+  scaleAxisIndex_ = scaleAxisIndex;
+  scaleTrans_ = scaleTrans;
+}
+
+void WidgetBase::setToScale(MouseButton btn, const Vector &flipAxis) {
+  if (btn == MouseButton::NONE) return;
+  movementType_[btn] = WidgetMovement::SCALE;
+  flipAxis_ = glm::vec3(flipAxis.x(), flipAxis.y(), flipAxis.z());
+}
+
+void WidgetBase::setToRotate(MouseButton btn)
 {
-  movementType_ = WidgetMovement::ROTATE;
+  if (btn == MouseButton::NONE) return;
+  movementType_[btn] = WidgetMovement::ROTATE;
 }
 
-void WidgetBase::setToTranslate()
+void WidgetBase::setToTranslate(MouseButton btn)
 {
-  movementType_ = WidgetMovement::TRANSLATE;
+  if (btn == MouseButton::NONE) return;
+  movementType_[btn] = WidgetMovement::TRANSLATE;
 }
 
 glm::vec3 WidgetBase::getFlipVector()
 {
   return flipAxis_;
 }
 
-WidgetMovement WidgetBase::getMovementType()
+int WidgetBase::getScaleAxisIndex()
+{
+  return scaleAxisIndex_;
+}
+
+glm::mat4* WidgetBase::getScaleTransform()
+{
+  return &scaleTrans_;
```

Comment:
Don't return pointers to internals. Use a reference instead to avoid copying, const reference if possible.